### DISCUSSION TITLE
Existing account key

### DISCUSSION THREAD

**Discussion by viharm** [go-acme/lego#2157] [Q&A] [answered]
Is it possible in LeGo to use an existing account key?

**Comment by ldez** [comment]
Hello,

technically, yes it's possible but not trivial, you have to reconstruct the following tree and files:

<details>
<summary>tree</summary>

```
.lego
└── accounts
    └── acme-staging-v02.api.letsencrypt.org
        └── myemail@exmaple.com
            ├── account.json
            └── keys
                └── myemail@exmaple.com.key

```

</details>

<details>
<summary>account.json</summary>

```json
{
	"email": "myemail@exmaple.com",
	"registration": {
		"body": {
			"status": "valid",
			"contact": [
				"mailto:myemail@exmaple.com"
			]
		},
		"uri": "https://acme-staging-v02.api.letsencrypt.org/acme/acct/123456789"
	}
}
```

</details>

`myemail@exmaple.com.key` is your private key.

**Reply by Eagle3386** [reply]
@ldez What if `lego` (as part of `certbot-multi-dns` which in turn is used in Home Assistant's "app" called [`letsencrypt`](https://github.com/home-assistant/addons/tree/master/letsencrypt)) runs in a Docker container, i.e., there's no chance of changing file contents via any given shell?

Can you offer a parameter (or allow for an environment variable) to override the account ID, i.e., the integer after `/acct/`?

**Reply by ldez** [reply]
I don't understand the need.

An "account" is several files, and must follow a specific file tree.
The private key is just one element.

As I don't understand your need, I cannot provide a solution.

**Reply by ldez** [reply]
After more thinking, I'm still not sure to understand the need, but I can create a command to "recover" an existing ACME server account from its private key.

**Reply by Eagle3386** [reply]
My bad, should've given more context. Here we go:
Using `lego` (via `certbot-multi-dns` on Home Assistant), another ACME account was created while I prefer to keep all certs under 1 account.

So, I basically like to have a parameter that lets me specify the account ID & `lego` than uses that for issuing, renewing, etc. That's because I don't have access to the files or the folder structure as its located within a container.

**Reply by ldez** [reply]
Could you open a dedicated discussion, and explain in detail your use-case?

**Comment by viharm** [comment]
Great, thanks. I'll try this.

**Comment by ldez** [comment, marked as answer]
In lego [v5](https://github.com/go-acme/lego/issues/2798) (unreleased for now), I added a new command `accounts recover` that allow importing an account by providing the private key.

**Reply by Eagle3386** [reply]
@ldez Great feature, but doesn't help with my issue. As you kindly requested last week, here's the dedicated discussion for it: https://github.com/go-acme/lego/discussions/2959

**Reply by Eagle3386** [reply]
Meanwhile, v5.x got released & I tried using this on a fresh download (`https://github.com/go-acme/lego/releases/download/v5.5.2/lego_v5.5.2_linux_amd64.tar.gz`) & extracted it to `/usr/local/share/lego` on my Synology NAS (716+II).
Using the account ID from the official URL, `accounturi=https://acme-v02.api.letsencrypt.org/acme/acct/12345678` (changed digits, but not their length), it keeps failing:
```sh
root@NAS:/usr/local/share/lego# ./lego accounts recover --account-id=12345678 --email user@example.net --key-type EC384 --private-key ../acme.sh/example.net_ecc/example.net.key --server letsencrypt
INFO  Recovering account. account=12345678 email=user@example.net server=https://acme-v02.api.letsencrypt.org/directory
Do you want to proceed? Y/n
y
INFO  Trying to resolve the account by key
WARN  retry: acme: error: 400 :: urn:ietf:params:acme:error:accountDoesNotExist :: No account exists with the provided key
ERROR Error error="resolve account by key: unable to communicate with the API server: error: Post \"https://acme-v02.api.letsencrypt.org/acme/new-acct\": POST https://acme-v02.api.letsencrypt.org/acme/new-acct giving up after 1 attempt(s): acme: error: 400 :: urn:ietf:params:acme:error:accountDoesNotExist :: No account exists with the provided key"
```
Omitting `--email` doesn't help, neither does omitting `--acount-id` (then relying on the email, of course).

However, the account _does_ exist as I just renewed my cert successfully. And the mentioned private key file is in PEM format as well.

**Reply by ldez** [reply]
`--acount-id` is not related to the ACME account ID, this is a string with a user defined name.

You are not using the command in the right way, so yes it doesn't work.

**Reply by Eagle3386** [reply]
Then where do I get the `account-id` that lego is expecting from? 🤔

**Reply by ldez** [reply]
This is the name of the directory that contains the account, and the value of the field `id` in the `account.json`.

```
./.lego/accounts/localhost_14000/foo@example.com/account.json
     │      │             │             └── accountID
     │      │             └── CA server ("server" option)
     │      └── root accounts directory
     └── "path" option
```

```json
{
  "id": "foo@example.com",
  "email": "lego@example.com",
  "keyType": "EC256",
  "server": "",
  "origin": "migration",
  "registration": {
    "status": "valid",
    "contact": [
      "mailto:lego@example.com"
    ],
    "orders": "https://localhost:14000/list-orderz/7cb7b237297d3f4d",
    "accountURL": "https://localhost:14000/my-account/7cb7b237297d3f4d"
  }
}

**Reply by ldez** [reply]
After reading again your previous message, in addition of the wrong assumption about the account ID you are using the certificate private key instead of the account private key.

**Reply by Eagle3386** [reply]
My bad, should've explain it better right from the beginning: I'm trying to import from ACME, so there's no `./.lego/accounts/localhost_14000/foo@example.com/account.json` folder path existing.

I naively thought all implementations are bound to adhere to ACME specifications, hence importing works across different clients, at least from `ACME.sh` to `le-go`, that is.

**Reply by ldez** [reply]
The command `accounts recover` can be used to import any account, independently of the client.

It imports account not certificates, so it need the private account key (not the private certificate key) to work.

The `account-id` is here to identify the account locally only.
You don't need to have the lego storage path because it will be created for you by using the `account-id`.

The ACME RFCs define the ways to do challenges, not how to store information, this is basically a server API documentation.
